system/default_ad_api/src/vehicle.cpp

```diff
@@ -0,0 +1,178 @@
+  adaptor.init_pub(pub_kinematics_);
+  adaptor.init_pub(pub_status_);
+  adaptor.init_sub(sub_kinematic_state_, this, &VehicleNode::kinematic_state);
+  adaptor.init_sub(sub_acceleration_, this, &VehicleNode::acceleration_status);
+  adaptor.init_sub(sub_steering_, this, &VehicleNode::steering_status);
+  adaptor.init_sub(sub_gear_state_, this, &VehicleNode::gear_status);
+  adaptor.init_sub(sub_turn_indicator_, this, &VehicleNode::turn_indicator_status);
+  adaptor.init_sub(sub_map_projector_info_, this, &VehicleNode::map_projector_info);
+  adaptor.init_sub(sub_hazard_light_, this, &VehicleNode::hazard_light_status);
+  adaptor.init_sub(sub_energy_level_, this, &VehicleNode::energy_status);
+
+  const auto rate = rclcpp::Rate(10);
+  timer_ = rclcpp::create_timer(this, get_clock(), rate.period(), [this]() { on_timer(); });
+}
+
+uint8_t VehicleNode::mapping(
+  std::unordered_map<uint8_t, uint8_t> hash_map, uint8_t input, uint8_t default_value)
+{
+  if (hash_map.find(input) == hash_map.end()) {
+    return default_value;
+  } else {
+    return hash_map[input];
+  }
+}
+
+void VehicleNode::kinematic_state(
+  const localization_interface::KinematicState::Message::ConstSharedPtr msg_ptr)
+{
+  kinematic_state_msgs_ = msg_ptr;
+}
+
+Eigen::Vector3d VehicleNode::toBasicPoint3dPt(const geometry_msgs::msg::Point src)
+{
+  Eigen::Vector3d dst;
+  dst.x() = src.x;
+  dst.y() = src.y;
+  dst.z() = src.z;
+  return dst;
+}
+
+void VehicleNode::acceleration_status(
+  const localization_interface::Acceleration::Message::ConstSharedPtr msg_ptr)
+{
+  acceleration_msgs_ = msg_ptr;
+}
+
+void VehicleNode::steering_status(
+  const vehicle_interface::SteeringStatus::Message::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.steering_tire_angle = msg_ptr->steering_tire_angle;
+}
+
+void VehicleNode::gear_status(const GearReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.gear.status = mapping(gear_type_, msg_ptr->report, ApiGear::UNKNOWN);
+}
+
+void VehicleNode::turn_indicator_status(const TurnIndicatorsReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.turn_indicators.status =
+    mapping(turn_indicator_type_, msg_ptr->report, ApiTurnIndicator::UNKNOWN);
+}
+
+void VehicleNode::hazard_light_status(const HazardLightsReport::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.hazard_lights.status =
+    mapping(hazard_light_type_, msg_ptr->report, ApiHazardLight::UNKNOWN);
+}
+
+void VehicleNode::map_projector_info(const MapProjectorInfo::ConstSharedPtr msg_ptr)
+{
+  map_projector_info_ = msg_ptr;
+}
+
+void VehicleNode::energy_status(
+  const vehicle_interface::EnergyStatus::Message::ConstSharedPtr msg_ptr)
+{
+  vehicle_status_.energy_percentage = msg_ptr->energy_level;
+}
+
+void VehicleNode::on_timer()
+{
+  vehicle_status_.stamp = now();
+  autoware_ad_api::vehicle::VehicleKinematics::Message vehicle_kinematics;
+  vehicle_kinematics.pose.header = kinematic_state_msgs_->header;
```

Comment:
It needs null check for `kinematic_state_msgs_` and `acceleration_msgs_`.

Reply:
@isamu-takagi I think the default will be 0, can you advice me how to do the null check?

Reply:
Something like:
```cpp
void VehicleNode::on_timer()
{
  publish_kinematics();
  publish_status();
}

void VehicleNode::publish_kinematics()
{
  if (!kinematic_state_msgs_ || !acceleration_msgs_) return;

  autoware_ad_api::vehicle::VehicleKinematics::Message vehicle_kinematics;
  vehicle_kinematics.pose.header = kinematic_state_msgs_->header;
  ...
}

void VehicleNode::publish_status()
{
  vehicle_status_.stamp = now();
  ...
}
```

It looks like the process died at `kinematic_state_msgs_->header`.

Reply:
fixed in [39bd585](https://github.com/autowarefoundation/autoware.universe/pull/2930/commits/39bd5852c76c25fa787c49096bb2f4dac98f2bfb)